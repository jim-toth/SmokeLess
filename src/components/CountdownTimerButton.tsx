/// <reference path="node_modules/typescript/lib/lib.d.ts" /
import * as React from 'react';
// import ReactMixin from 'react-mixin';
// import * as TimerMixin from 'react-timer-mixin';
import { Button } from 'react-native';

// Some code modified from https://github.com/noelyoo/react-native-timer-countdown/blob/master/index.tsx

interface ICountdownTimerButtonProps {
  bottomText?: string;
  until?: Date;
  resetOnPress?: boolean;
  onPress?: (beforeTimerExpired: boolean) => void;
}

interface ICountdownTimerButtonState {
  millisecondsRemaining: number,
  timeoutId: any,
  previousMilliseconds?: number,
  isExpired: boolean
}

export default class CountdownTimerButton extends React.Component<ICountdownTimerButtonProps, ICountdownTimerButtonState> {
  public readonly state = {
    millisecondsRemaining: this.props.until ? this.props.until.getTime() - Date.now() : -1,
    timeoutId: 0,
    previousMilliseconds: 0,
    isExpired: false
  }

  public componentDidMount(): void {
    this.tick();
  }

  public componentWillReceiveProps(newProps:ICountdownTimerButtonProps): void {
    if (this.state.timeoutId !== undefined) {
      clearTimeout(this.state.timeoutId);
    }
    this.setState({
      previousMilliseconds: undefined,
      millisecondsRemaining: newProps.until ? newProps.until.getTime() - Date.now() : -1
    });
  }

  public componentDidUpdate(): void {
    if (!this.state.previousMilliseconds && this.state.millisecondsRemaining > 0) {
      this.tick();
    }
  }

  public componentWillUnmount(): void {
    clearTimeout(this.state.timeoutId);
  }

  private isExpired = () : boolean => {
    return this.state.isExpired;
  }

  _onPress = () => {
    if (this.props.onPress) {
      this.props.onPress(this.isExpired());
    }
  }

  private tick = () => {
    const currentMilliseconds = Date.now();
    const dt = this.state.previousMilliseconds
      ? currentMilliseconds - this.state.previousMilliseconds
      : 0;
    const interval: number = 1000;

    // correct for small variations in actual timeout time
    const intervalSecondsRemaing: number = interval - (dt % interval);
    let timeout: number = intervalSecondsRemaing;

    if (intervalSecondsRemaing < interval / 2.0) {
      timeout += interval;
    }

    const millisecondsRemaining: number = Math.max(this.state.millisecondsRemaining - dt, 0);
    const isComplete: boolean = this.state.previousMilliseconds && millisecondsRemaining <= 0 ? true : false;

    if (this.state.timeoutId !== undefined) {
      clearTimeout(this.state.timeoutId);
    }

    this.setState({
      timeoutId: isComplete ? undefined : setTimeout(this.tick, timeout),
      previousMilliseconds: currentMilliseconds,
      millisecondsRemaining,
      isExpired: isComplete
    });

    // if (this.props.onTick !== undefined) {
    //   this.props.onTick(millisecondsRemaining);
    // }
  };

  private getFormattedTime = (milliseconds: number): string => {
    if (milliseconds < 0) return '--:--:--'
    // if (this.props.formatMilliseconds !== undefined) {
    //   return this.props.formatMilliseconds(milliseconds);
    // }
    const remainingSec: number = Math.round(milliseconds / 1000);

    const seconds: number = parseInt((remainingSec % 60).toString(), 10);
    const minutes: number = parseInt(((remainingSec / 60) % 60).toString(), 10);
    const hours: number = parseInt((remainingSec / 3600).toString(), 10);

    const s = seconds < 10 ? '0' + seconds : seconds;
    const m = minutes < 10 ? '0' + minutes : minutes;
    let h = hours < 10 ? '0' + hours : hours;
    h = h === '00' ? '' : h + ':';
    return h + m + ':' + s;
  };

  render() {
    const millisecondsRemaining: number = this.state.millisecondsRemaining;
    const buttonColor = millisecondsRemaining > 0 ? 'red' : 'green';

    return (
      <Button
        color={buttonColor}
        title={this.getFormattedTime(millisecondsRemaining)}
        onPress={this._onPress}>
      </Button>
    );
  }
}

// ReactMixin(CountdownTimerButton.prototype, TimerMixin);
