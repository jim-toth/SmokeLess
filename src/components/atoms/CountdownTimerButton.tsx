/// <reference path="node_modules/typescript/lib/lib.d.ts" /
import * as React from 'react';
// import ReactMixin from 'react-mixin';
// import * as TimerMixin from 'react-timer-mixin';
import { View, Text, TouchableNativeFeedback } from 'react-native';

// Some code modified from https://github.com/noelyoo/react-native-timer-countdown/blob/master/index.tsx

const defaultStyle:any = {
  button: {
    width: '80%',
    backgroundColor: 'red',
    borderWidth: 2,
    borderColor: 'red',
    borderRadius: 4
  },
  buttonExpired: {
    width: '80%',
    backgroundColor: 'green',
    borderWidth: 2,
    borderColor: 'green',
    borderRadius: 4
  },
  text: {
    color: 'white',
    margin: 5,
    width: '100%',
    textAlign: 'center',
    fontFamily: 'saira'
  }
}

interface ICountdownTimerButtonProps {
  bottomText?: string;
  until?: Date;
  resetOnPress?: boolean;
  onPress?: (beforeTimerExpired: boolean) => void;
  buttonStyle?: any;
  buttonExpiredStyle?: any;
  textStyle?: any;
}

interface ICountdownTimerButtonState {
  millisecondsRemaining: number,
  timeoutId: any,
  previousMilliseconds?: number
}

export default class CountdownTimerButton extends React.Component<ICountdownTimerButtonProps, ICountdownTimerButtonState> {
  constructor(props:ICountdownTimerButtonProps) {
    super(props);

    const millisecondsRemaining = props.until ? props.until.getTime() - Date.now() : -1;

    this.state = {
      millisecondsRemaining,
      timeoutId: 0,
      previousMilliseconds: 0
    }
  }

  static defaultProps = {
    buttonStyle: defaultStyle.button,
    buttonExpiredStyle: defaultStyle.buttonExpired,
    textStyle: defaultStyle.text
  }

  public componentDidMount(): void {
    this.tick();
  }

  public componentWillReceiveProps(newProps:ICountdownTimerButtonProps): void {
    if (this.state.timeoutId !== undefined) {
      clearTimeout(this.state.timeoutId);
    }
    this.setState({
      previousMilliseconds: undefined,
      millisecondsRemaining: newProps.until ? newProps.until.getTime() - Date.now() : -1
    });
  }

  public componentDidUpdate(): void {
    if (!this.state.previousMilliseconds && this.state.millisecondsRemaining > 0) {
      this.tick();
    }
  }

  public componentWillUnmount(): void {
    clearTimeout(this.state.timeoutId);
  }

  private isExpired = () : boolean => {
    return this.state.millisecondsRemaining <= 0;
  }

  _onPress = () => {
    if (this.props.onPress) {
      this.props.onPress(this.isExpired());
    }
  }

  private tick = () => {
    const currentMilliseconds = Date.now();
    const dt = this.state.previousMilliseconds
      ? currentMilliseconds - this.state.previousMilliseconds
      : 0;
    const interval: number = 1000;

    // correct for small variations in actual timeout time
    const intervalSecondsRemaing: number = interval - (dt % interval);
    let timeout: number = intervalSecondsRemaing;

    if (intervalSecondsRemaing < interval / 2.0) {
      timeout += interval;
    }

    const millisecondsRemaining: number = Math.max(this.state.millisecondsRemaining - dt, 0);
    // const isComplete: boolean = this.state.previousMilliseconds && millisecondsRemaining <= 0 ? true : false;
    const isComplete: boolean = millisecondsRemaining <= 0 ? true : false;

    if (this.state.timeoutId !== undefined) {
      clearTimeout(this.state.timeoutId);
    }

    this.setState({
      timeoutId: isComplete ? undefined : setTimeout(this.tick, timeout),
      previousMilliseconds: currentMilliseconds,
      millisecondsRemaining
    });
  };

  private getFormattedTime = (milliseconds: number): string => {
    if (milliseconds <= 0) return '--:--:--'

    const remainingSec: number = Math.round(milliseconds / 1000);

    const seconds: number = parseInt((remainingSec % 60).toString(), 10);
    const minutes: number = parseInt(((remainingSec / 60) % 60).toString(), 10);
    const hours: number = parseInt((remainingSec / 3600).toString(), 10);

    const s = seconds < 10 ? '0' + seconds : seconds;
    const m = minutes < 10 ? '0' + minutes : minutes;
    let h = hours < 10 ? '0' + hours : hours;
    h = h === '00' ? '' : h + ':';
    return h + m + ':' + s;
  };

  render() {
    return (
      <TouchableNativeFeedback
        onPress={this._onPress}
        background={TouchableNativeFeedback.SelectableBackground()}
      >
        <View style={this.isExpired() ? this.props.buttonExpiredStyle : this.props.buttonStyle}>
          <Text style={this.props.textStyle}>{this.getFormattedTime(this.state.millisecondsRemaining)}</Text>
        </View>
      </TouchableNativeFeedback>
    );
  }
}
