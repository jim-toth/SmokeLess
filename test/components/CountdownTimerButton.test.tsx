import 'react-native';
import * as React from 'react';
import CountdownTimerButton from '../../src/components/CountdownTimerButton';
import { shallow } from 'enzyme';

describe('<CountdownTimerButton />', () => {
  let anHourFromNow = new Date();

  beforeAll(() => {
    anHourFromNow.setHours(anHourFromNow.getHours() + 1);
  });

  it('renders duration on a button from now until expiration time', () => {
    const tree = shallow(
      <CountdownTimerButton
        until={anHourFromNow}
      />
    );

    expect(tree.find('Button')).toHaveLength(1);
    expect(tree.find('Button').first().prop('title')).toEqual('01:00:00');
  });

  it('counts down', () => {
    const tree = shallow(
      <CountdownTimerButton
        until={anHourFromNow}
      />
    );

    jest.runOnlyPendingTimers();

    expect(tree.find('Button').first().prop('title')).toEqual('59:59');
  });

  it('matches snapshot', () => {
    const tree = shallow(
      <CountdownTimerButton
        until={new Date('1996-12-25')}
      />
    );

    expect(tree).toMatchSnapshot();
  });
});