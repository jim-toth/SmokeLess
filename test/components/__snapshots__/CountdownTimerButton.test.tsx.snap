// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CountdownTimerButton /> matches snapshot 1`] = `
<Button
  onPress={[Function]}
  title="00:00"
/>
`;
